### PR TITLE
444 split runtime initialization

### PR DESCRIPTION
**Pull request by uhetmaniuk**
Split the runtime initialization.
Allow for a third-party specification of arguments.
Set precedence rule for each argument.
Generate the startup banner.
Give the option to create a ".INI" file with the simulation parameters.

### CODE REVIEW

**Comment by uhetmaniuk** [comment]
@lifflander The management of the anchors is not as fancy as your suggestion in https://github.com/DARMA-tasking/vt/tree/444-arg-precedence-resolution
But I believe it has the same functionalities.
Let me know if I miss anything.

**Comment by uhetmaniuk** [comment]
@lifflander Implemented the printing as requested. The code is ready to be reviewed.

One question: do we need the flag "vt_color"? It can not be turned off from the command line and the flag "vt_no_color" actually does it. At first sight, it looks redundant to me.

File: src/vt/runtime/runtime.cc [hunk 1]
```diff
@@ -701,6 +729,11 @@ void Runtime::printShutdownBanner(term::TermCounterType const& num_units) {
 }
 
 bool Runtime::initialize(bool const force_now) {
+  //--- Test whether (command-line, input file) arguments
+  //--- have been parsed.
+  if (!parsed_arg)
+    return false;
+  //
```
Comment (pnstickne) [inline src/vt/runtime/runtime.cc]: Could move comment to member variable itself and/or expand variable name.

    // True if argument parsing (from CLI or input file) has been performed
    bool arguments_parsed;

File: src/vt/runtime/runtime.h [hunk 2]
```diff
@@ -175,6 +191,7 @@ struct Runtime {
   MPI_Comm* communicator_ = nullptr;
   int user_argc_ = 0;
   char** user_argv_ = nullptr;
+  bool parsed_arg = false;
```
Comment (pnstickne) [inline src/vt/runtime/runtime.h]: Perhaps `parsed_arg_`, and private if allowable.

File: src/vt/runtime/runtime.cc [hunk 3]
```diff
@@ -96,12 +96,40 @@ Runtime::Runtime(
...
 }
 
+Runtime::Runtime(
+  bool const interop_mode, 
+  RuntimeInstType const in_instance
+)  : instance_(in_instance), runtime_active_(false), is_interop_(interop_mode),
+     num_workers_(no_workers), communicator_(nullptr), user_argc_(0),
+     user_argv_(nullptr)
+{
+  sig_user_1_ = false;
```
Comment (pnstickne) [inline src/vt/runtime/runtime.cc]: Use argument (?) initialization for sig_user_1_ too? What about the other members? Seems like should consistent between..

File: src/vt/configs/arguments/args.cc [hunk 4]
```diff
@@ -194,22 +127,31 @@ namespace vt { namespace arguments {
...
-  auto m = app.add_option("--vt_stack_mod",     vt_stack_mod,       mod,  1);
-  auto stackGroup = "Dump Stack Backtrace";
-  g->group(stackGroup);
-  h->group(stackGroup);
-  i->group(stackGroup);
-  j->group(stackGroup);
-  k->group(stackGroup);
-  l->group(stackGroup);
-  m->group(stackGroup);
+  auto g = setup_.addFlag("vt_no_warn_stack",   config.vt_no_warn_stack,   warn);
```
Comment (pnstickne) [inline src/vt/configs/arguments/args.cc]: Variable names seem a bit too terse :|

File: src/vt/configs/arguments/args.cc [hunk 5]
```diff
@@ -414,37 +264,1238 @@ namespace vt { namespace arguments {
...
+  u2->setGroup(userOpts);
+  u3->setGroup(userOpts);
+  ui1->setGroup(userOpts);
+  ui2->setGroup(userOpts);
+  ui3->setGroup(userOpts);
+  us1->setGroup(userOpts);
+  us2->setGroup(userOpts);
+  us3->setGroup(userOpts);
+
+  //---
```
Comment (pnstickne) [inline src/vt/configs/arguments/args.cc]: Separator '//---' comments and newline spaces maybe reducible.

File: src/vt/configs/arguments/args.cc [hunk 6]
```diff
@@ -414,37 +264,1238 @@ namespace vt { namespace arguments {
...
+   */
+
+  std::unordered_map<config::CatEnum, std::tuple<bool*, std::string, std::string> > debug_flag;
+
+  debug_flag[config::none] = std::make_tuple<bool*, std::string, std::string>(
+         &config.vt_debug_none, 
+         std::string("debug_none"), 
+		 std::string(config::PrettyPrintCat<config::none>::str) 
+		 );
+  debug_flag[config::gen]  = std::make_tuple(
```
Comment (pnstickne) [inline src/vt/configs/arguments/args.cc]: Seems like could be reduced to common function or template function.

File: src/vt/configs/arguments/args.cc [hunk 7]
```diff
@@ -414,37 +264,1238 @@ namespace vt { namespace arguments {
...
+
+  for (const auto dcase : debug_flag) {
+    auto f3 = dcase.second;
+	auto ff = std::string("Enable ") + std::get<1>(f3) 
+		+ std::string(" = \"") + std::get<2>(f3) + std::string("\"");
+	auto &fbool = *(std::get<0>(f3));
+	auto fname = std::string("vt_") + std::get<1>(f3);
+	auto fr = setup_.addFlag(fname, fbool, ff);
+	fr->setGroup(debugGroup);
+	if (!((vt::config::DefaultConfig::category & dcase.first) != 0)) {
```
Comment (pnstickne) [inline src/vt/configs/arguments/args.cc]: !((vt::config::DefaultConfig::category & dcase.first) != 0)

Maybe just: `(vt::config::DefaultConfig::category & dcase.first) == 0`

File: src/vt/configs/arguments/args.cc [hunk 8]
```diff
@@ -414,37 +264,1238 @@ namespace vt { namespace arguments {
...
+
+  auto r1 = setup_.addFlag("vt_debug_verbose", config.vt_debug_verbose,
+		  "Enable verbose debug prints");
+  r1->setGroup(debugGroup);
+
+  for (const auto dcase : debug_flag) {
+    auto f3 = dcase.second;
+	auto ff = std::string("Enable ") + std::get<1>(f3) 
+		+ std::string(" = \"") + std::get<2>(f3) + std::string("\"");
+	auto &fbool = *(std::get<0>(f3));
```
Comment (pnstickne) [inline src/vt/configs/arguments/args.cc]: What does `fbool` mean, semantically?

File: src/vt/configs/arguments/args.cc [hunk 9]
```diff
@@ -414,37 +264,1238 @@ namespace vt { namespace arguments {
...
+	auto fname = std::string("vt_") + std::get<1>(f3);
+	auto fr = setup_.addFlag(fname, fbool, ff);
+	fr->setGroup(debugGroup);
+	if (!((vt::config::DefaultConfig::category & dcase.first) != 0)) {
+      if (fbool)
+	    fr->setBannerMsg_Warning(std::get<1>(f3));
```
Comment (pnstickne) [inline src/vt/configs/arguments/args.cc]: Already using `get<1>(f3)` above, a local name may make usage more clear. I suppose it's a "message" of some kind.

File: src/vt/configs/arguments/args.cc [hunk 10]
```diff
@@ -414,37 +264,1238 @@ namespace vt { namespace arguments {
...
+  }
+
+}
+
+/* ------------------------------------------------- */
+//--- Utility functions
+/* ------------------------------------------------- */
+
+template< typename T>
+std::string display(const T &val) { std::string res; return res; }
```
Comment (pnstickne) [inline src/vt/configs/arguments/args.cc]: Name sounds like action, although they are side-effect free functions. 'getDisplayValue'?

File: src/vt/configs/arguments/args.cc [hunk 11]
```diff
@@ -194,22 +127,31 @@ namespace vt { namespace arguments {
...
+  l->setGroup(stackGroup);
+  m->setGroup(stackGroup);
+
+  //--- Indicates that true 'j' skips the options g, h, i
+  g->needsOptionOff(j);
+  h->needsOptionOff(j);
+  i->needsOptionOff(j);
 
+  //--- Indicates that options k, l, m are used when j is 'false'
+  k->needsOptionOff(j);
```
Comment (pnstickne) [inline src/vt/configs/arguments/args.cc]: Why is the comment different when the method called is the same?

File: src/vt/configs/arguments/args.cc [hunk 12]
```diff
@@ -414,37 +264,1238 @@ namespace vt { namespace arguments {
...
+  fmt::print("{}\t{}{}", vt_pre, f9, reset);
+
+}
+
+/* ------------------------------------------------- */
+// --- Member functions for class Anchor<T>
+/* ------------------------------------------------- */
+
+
+template< typename T >
```
Comment (pnstickne) [inline src/vt/configs/arguments/args.cc]: In layman terms, what is an 'Anchor'? It's not a term I'd normally associate with parameters..

File: src/vt/configs/arguments/args.cc [hunk 13]
```diff
@@ -414,37 +264,1238 @@ namespace vt { namespace arguments {
...
+    auto goptions = getGroupOptions(gname);
+    for (const auto &opt : goptions) {
+      opt.second->print();
+    }
+  }
+}
+
+//-----
+
+void ArgSetup::setup_Printing()
```
Comment (pnstickne) [inline src/vt/configs/arguments/args.cc]: Names could be more consistent.

File: src/vt/configs/arguments/args.cc [hunk 14]
```diff
@@ -414,37 +264,1238 @@ namespace vt { namespace arguments {
...
+  b->setBannerMsg_OnOff("Disabling color output", "Color output enabled by default");
+  c->setBannerMsg_On("Automatic TTY detection for color output",
+        [&]() { return (Args::config.vt_no_color == false); });
+
+  auto quiet  = "Quiet the output from vt (only errors, warnings)";
+  a1->setBannerMsg_On(quiet);
+
+  //----
+
+  auto d = getOption<bool>("vt_no_SIGINT");
```
Comment (pnstickne) [inline src/vt/configs/arguments/args.cc]: Maybe minimal names:

    auto optINT = ..
    auto optSEGV = ..
    auto optTerm = ..

    auto optQuiet = ..
    optQuiet->setBannerMsgOn(quiet);

File: src/vt/configs/arguments/args.cc [hunk 15]
```diff
@@ -414,37 +264,1238 @@ namespace vt { namespace arguments {
...
+
+int ArgSetup::parse(int& argc, char**& argv)
+{
+
+  if (parsed)
+    return 0;
+
+  // CLI11 app parser expects to get the arguments in *reverse* order!
+  std::vector<std::string> args;
+  for (auto i = argc-1; i > 0; i--) {
```
Comment (pnstickne) [inline src/vt/configs/arguments/args.cc]: `int` - auto not adding much value in context.

File: src/vt/configs/arguments/args.cc [hunk 16]
```diff
@@ -470,15 +1521,18 @@ namespace vt { namespace arguments {
...
   }
 
   // Set them back with all vt arguments elided
   argc = new_argc;
   argv = new_argv;
 
   parsed = true;
-  return 1;
+
+  return 0;
```
Comment (pnstickne) [inline src/vt/configs/arguments/args.cc]: Return 0 now vs 1 before?

Comment (pnstickne) [reply src/vt/configs/arguments/args.cc]: (Does return value have any meaning anymore?)

File: src/vt/configs/arguments/args.cc (hunk 12 above)
Comment (uhetmaniuk) [reply src/vt/configs/arguments/args.cc]: @lifflander suggested those names: AnchorBase (for the virtual base class) and Anchor<T> for the actual instance.

File: src/vt/configs/arguments/args.cc (hunk 16 above)
Comment (uhetmaniuk) [reply src/vt/configs/arguments/args.cc]: OK. Will update.

File: src/vt/configs/arguments/args.cc (hunk 15 above)
Comment (uhetmaniuk) [reply src/vt/configs/arguments/args.cc]: OK. Will do.

File: src/vt/runtime/runtime.h (hunk 2 above)
Comment (uhetmaniuk) [reply src/vt/runtime/runtime.h]: OK.

File: src/vt/runtime/runtime.cc (hunk 1 above)
Comment (uhetmaniuk) [reply src/vt/runtime/runtime.cc]: OK.

File: src/vt/configs/arguments/args.cc (hunk 5 above)
Comment (uhetmaniuk) [reply src/vt/configs/arguments/args.cc]: OK.

File: src/vt/configs/arguments/args.cc (hunk 10 above)
Comment (uhetmaniuk) [reply src/vt/configs/arguments/args.cc]: OK

File: src/vt/configs/arguments/args.cc (hunk 11 above)
Comment (uhetmaniuk) [reply src/vt/configs/arguments/args.cc]: OK

File: src/vt/configs/arguments/args.cc (hunk 13 above)
Comment (uhetmaniuk) [reply src/vt/configs/arguments/args.cc]: OK

File: src/vt/configs/arguments/args.cc [hunk 17]
```diff
@@ -415,37 +259,1195 @@ namespace vt { namespace arguments {
...
+//--- Utility printing classes
+/* ------------------------------------------------- */
+
+class Printer {
+public:
+  virtual void output() = 0;
+};
+
+template < typename T >
+class Print_On : public Printer {
```
Comment (lifflander) [inline src/vt/configs/arguments/args.cc]: Please use camel case on struct names... as specified in the VT style guidelines

File: src/vt/configs/arguments/args.cc (hunk 12 above)
Comment (lifflander) [reply src/vt/configs/arguments/args.cc]: An `Anchor` is an specific central argument resolution point that might have different names in various levels of configuration. I'm open to a better name (for that and `AnchorBase`).

File: src/vt/configs/arguments/args.cc [hunk 18]
```diff
@@ -415,37 +259,1195 @@ namespace vt { namespace arguments {
...
+
+  debug_flag[config::none] = std::make_tuple<bool*, std::string, std::string>(
+         &config.vt_debug_none, 
+         std::string("debug_none"), 
+		 std::string(config::PrettyPrintCat<config::none>::str) 
+		 );
+  debug_flag[config::gen]  = std::make_tuple(
+		 &config.vt_debug_gen,
+         std::string("debug_gen"), 
+		 std::string(config::PrettyPrintCat<config::gen>::str)
```
Comment (lifflander) [inline src/vt/configs/arguments/args.cc]: Indentation on all these lines looks a bit wonky. Are you using tabs?

File: src/vt/configs/arguments/args.cc [hunk 19]
```diff
@@ -415,37 +259,1195 @@ namespace vt { namespace arguments {
...
+	}
+  }
+
+}
+
+/* ------------------------------------------------- */
+//--- Utility functions
+/* ------------------------------------------------- */
+
+template< typename T>
```
Comment (lifflander) [inline src/vt/configs/arguments/args.cc]: Can you split these into a argument utility header file?

File: src/vt/configs/arguments/args.cc [hunk 20]
```diff
@@ -415,37 +259,1195 @@ namespace vt { namespace arguments {
...
+  auto wy = getOption<std::string>("vt_lb_stats_file");
+
+#if !backend_check_enabled(lblite)
+  s->setBannerMsg_Warning("lblite");
+  ww->setBannerMsg_Warning("lblite");
+#else
+  s->setBannerMsg_On("Load balancing enabled");
+  {
+	auto lb_quiet      = "Silence load balancing output";
+	t1->setBannerMsg_On(lb_quiet,
```
Comment (lifflander) [inline src/vt/configs/arguments/args.cc]: Indentation looks off.

File: src/vt/configs/arguments/args.h [hunk 21]
```diff
@@ -45,106 +45,678 @@
...
+   /// \param[in] msg_on String to display when the anchor is active
+   /// \param[in] msg_off String to display when the anchor is off
+   /// \param[in] fun Additional condition to print or not the message
+   ///
+   /// \note The ON message will be of the form
+   /// "Option: flag 'NAME' on: 'MSG_ON'"
+   /// \note The OFF message will be of the form
+   /// "Default: 'MSG_OFF', use 'NAME' to disable"
+   ///
+   void setBannerMsg_OnOff(std::string msg_on, std::string msg_off,
```
Comment (lifflander) [inline src/vt/configs/arguments/args.h]: Use consistent camel case for these, please.

File: src/vt/trace/trace.h [hunk 22]
```diff
@@ -89,29 +83,24 @@ struct Trace {
...
 
   void beginProcessing(
     TraceEntryIDType const ep, TraceMsgLenType const len,
     TraceEventIDType const event, NodeType const from_node,
-    double const time = getCurrentTime(),
-    uint64_t const idx1 = 0, uint64_t const idx2 = 0, uint64_t const idx3 = 0,
-    uint64_t const idx4 = 0
-  );
+    double const time = getCurrentTime(), uint64_t const idx1 = 0,
+    uint64_t const idx2 = 0, uint64_t const idx3 = 0, uint64_t const idx4 = 0);
```
Comment (nlslatt) [inline src/vt/trace/trace.h]: Many of the changes made in the most recent commit are against the vt coding guidelines.  The original line wrapping here was correct.

File: src/vt/configs/arguments/args.cc (hunk 17 above)
Comment (uhetmaniuk) [reply src/vt/configs/arguments/args.cc]: OK. Updated.

File: src/vt/configs/arguments/args.cc (hunk 18 above)
Comment (uhetmaniuk) [reply src/vt/configs/arguments/args.cc]: OK. Used clang-format to check the formatting.

File: src/vt/configs/arguments/args.cc (hunk 20 above)
Comment (uhetmaniuk) [reply src/vt/configs/arguments/args.cc]: OK. Used clang-format to clean up the indentations.

File: src/vt/configs/arguments/args.h (hunk 21 above)
Comment (uhetmaniuk) [reply src/vt/configs/arguments/args.h]: OK. Cleaned up the names.

File: src/vt/trace/trace.h (hunk 22 above)
Comment (uhetmaniuk) [reply src/vt/trace/trace.h]: OK. Ran clang-format on the files. It should hopefully fixed most formatting issues.

Comment (lifflander) [reply src/vt/trace/trace.h]: Clang-format approximates this. The full guidelines are specified in this file: [https://github.com/DARMA-tasking/vt/blob/develop/style-guidelines.txt](https://github.com/DARMA-tasking/vt/blob/develop/style-guidelines.txt)

**Comment by lifflander** [comment]
Clang-format approximates the style. The full guidelines are specified in this file: [https://github.com/DARMA-tasking/vt/blob/develop/style-guidelines.txt](https://github.com/DARMA-tasking/vt/blob/develop/style-guidelines.txt)

File: src/vt/configs/arguments/args.cc (hunk 19 above)
Comment (uhetmaniuk) [reply src/vt/configs/arguments/args.cc]: OK. Done.

File: src/vt/configs/arguments/args.cc (hunk 14 above)
Comment (uhetmaniuk) [reply src/vt/configs/arguments/args.cc]: Done.

File: src/vt/configs/arguments/args.cc (hunk 6 above)
Comment (uhetmaniuk) [reply src/vt/configs/arguments/args.cc]: I am not sure how to address that point. I am open to suggestions.

File: src/vt/collective/collective_ops.cc [hunk 23]
```diff
@@ -81,6 +81,36 @@ RuntimePtrType CollectiveAnyOps<instance>::initialize(
...
+  bool is_interop,
+  WorkerCountType const num_workers,
+  MPI_Comm* comm
+) {
+  using vt::runtime::RuntimeInst;
+  using vt::runtime::Runtime;
+  using vt::runtime::eRuntimeInstance;
+
+#pragma sst global rt
+  RuntimeInst<instance>::rt = std::make_unique<Runtime>(is_interop);
```
[the remark this answers is not in the document]
Comment (uhetmaniuk) [reply src/vt/collective/collective_ops.cc]: allocate(...) is based on the initialize(...) routine, hence the multiple pragmas.
I am not sure how to address the comment. Any suggestion?

File: src/vt/runtime/runtime.cc (hunk 3 above)
Comment (uhetmaniuk) [reply src/vt/runtime/runtime.cc]: sig_user_1_ is static and has been initialized outside the constructor.
This new constructor is extracted from the first one, hence the similar initialization list.

**Comment by uhetmaniuk** [comment]
@lifflander How can I see where Docker is failing? (The link "Details" asks for a user-id + password.)

**Comment by lifflander** [comment]
I know this has been sitting for a long time. I want to get this on develop if we can. I'll do a full review, test, and re-base soon.

**Comment by uhetmaniuk** [comment]
@lifflander 
I fear that the rebase will be messy (as I may have merged with develop during the progress). 
I will try to create a new branch (based on the latest develop) to simplify your review/rebase.
How will this PR relate to the newer work #607 ?

**Comment by lifflander** [comment]
Good question. Maybe we should hold off until we reach a resolution with #607?

**Comment by uhetmaniuk** [comment]
I will at least do some clean up for a "simpler" future rebase.

**Comment by lifflander** [comment]
Ok, thanks!

File: src/vt/collective/collective_ops.cc (hunk 23 above)
[the remark this answers is not in the document]
Comment (pnstickne) [reply src/vt/collective/collective_ops.cc]: Instead of updating the global initially, use a local variable. It would still be a unique_ptr to the object - so there would be no unexpected copy operators.

Then assign to the global once at the end of all the operations. With a change like this there would be at most one `#pragma sst global rt` because all the other accesses would be local.

[the remark this answers is not in the document]
Comment (pnstickne) [reply src/vt/collective/collective_ops.cc]: This would also probably allow simplification of `rt_ptr` (ie. use the unique_ptr all the way through the function).

[the remark this answers is not in the document]
Comment (pnstickne) [reply src/vt/collective/collective_ops.cc]: What is the end goal of `CollectiveOps`? From my understanding, it should generally be migrated away from..

[the remark this answers is not in the document]
Comment (pnstickne) [reply src/vt/collective/collective_ops.cc]: Also, does the ordering of MPI_Init change? That's the real current 'MPI_Init violation' in RT, made difficult to expose with side-effects-of-ctors..

File: src/vt/collective/collective_ops.h [hunk 24]
```diff
@@ -66,6 +66,17 @@ struct CollectiveAnyOps {
     int& argc, char**& argv, WorkerCountType const num_workers = no_workers,
     bool is_interop = false, MPI_Comm* comm = nullptr
   );
+
+  ///
+  /// \brief Allocate routine to allow for two-step initialization
```
Comment (pnstickne) [inline src/vt/collective/collective_ops.h]: `allocate` seems a bit strange of a name.

How would one use this in a program?

**Comment by uhetmaniuk** [comment]
The history for the branch is all mixed (due to merges with develop).
Closing and replacing with #646
